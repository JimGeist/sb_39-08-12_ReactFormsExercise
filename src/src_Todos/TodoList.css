.TodoList-Main {
    clear: both;
    padding-top: 2.0rem;
}