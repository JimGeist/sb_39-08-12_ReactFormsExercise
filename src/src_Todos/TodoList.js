import React, { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

import "./TodoList.css";

function TodoList() {

    const [todoList, setTodoList] = useState([])

    const addTodo = (todoData) => {
        setTodoList(
            [...todoList, { ...todoData }]
        );
    }

    const deleteTodo = (id) => (
        setTodoList(
            todoList.filter(todo => todo.id !== id)
        )
    )

    return (
        <div className="TodoList-Main">
            <h1>Todo List</h1>
            <NewTodoForm addTodoFx={addTodo} />
            <ul>
                {
                    todoList.map(({ id, todo }) =>
                        <Todo id={id} todo={todo} deleteFx={deleteTodo} key={id} />
                    )
                }
            </ul>
        </div >
    )

}

export default TodoList;