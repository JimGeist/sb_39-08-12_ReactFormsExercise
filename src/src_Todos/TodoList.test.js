import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

// Smoke Test -- does TodoList render?
it("renders the TodoList component without crashing", () => {
    render(<TodoList />);
});

// TodoList snapshot tests
it("TodoList component matches snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("has a 'New todo:' label and an 'Add New Todo' button", function () {
    const { getByLabelText, getByText } = render(<TodoList />);

    // expect 'New todo:' label
    expect(getByLabelText("New todo:")).toBeInTheDocument();

    // expect 'Add New Todo' button
    expect(getByText("Add New Todo")).toBeInTheDocument();

});

it("can add a new todo item and delete it", function () {
    const testTodoText = ">> this is a test todo item <<";

    const { queryByTestId, getByLabelText, queryByText } = render(<TodoList />);

    // todo list element does not exist
    expect(queryByTestId("^Todo Item^")).not.toBeInTheDocument();
    expect(queryByText(testTodoText)).not.toBeInTheDocument();

    const todoText = getByLabelText("New todo:");
    const todoAddBtn = queryByText("Add New Todo")

    // fill out the form
    fireEvent.change(todoText, { target: { value: testTodoText } });
    fireEvent.click(todoAddBtn);

    // to do list element now exists
    expect(queryByTestId("^Todo Item^")).toBeInTheDocument();
    expect(queryByText(testTodoText)).toBeInTheDocument();

    const todoDeleteBtn = queryByText("X");
    fireEvent.click(todoDeleteBtn);

    // box no longer exists
    expect(queryByTestId("^Todo Item^")).not.toBeInTheDocument();
    expect(queryByText(testTodoText)).not.toBeInTheDocument();

});
