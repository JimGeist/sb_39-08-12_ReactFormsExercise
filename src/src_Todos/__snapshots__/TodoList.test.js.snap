// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList-Main"
  >
    <h1>
      Todo List
    </h1>
    <form
      class="NewTodoForm"
    >
      <label
        for="todo"
      >
        New todo:
      </label>
      <input
        class="NewTodoForm-InpTodo"
        id="todo"
        name="todo"
        type="text"
        value=""
      />
      <button
        class="NewTodoForm-BtnAdd"
      >
        Add New Todo
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
